docs/blueprints/log-in-with-storj.md

```diff
@@ -0,0 +1,205 @@
+accounts in each region which presents a challenge when merging them into a single table.
+
+So how can we provide the desired developer experience from the second solution, but with the technical simplicity of
+the first solution? Something we might consider is how this picture looks with a three-legged oauth (commonly used by
+banking systems).
+
+![three-legged oauth](images/log-in-with-storj-three-legged-oauth.svg)
+
+In this pattern, the identity provider (or authentication server) is "separate" from the resources that it may be
+granting access to. Once the client is granted access, the resource server can validate the credential against the
+authentication server. To Alice, everything is appears as Storj is granting StorJIF access to their resources.
+
+Once a client completes the OAuth workflow, they can perform requests on behalf of the end user for as long as the token
+is valid. To Alice, they're just interacting with this new StorJIF application, but they own all their data.
+
+![common request flow](images/log-in-with-storj-request.svg)
+
+## Rationale
+
+_Why OpenID Connect?_
+
+OpenID Connect has become a fairly strong (and consistent) standard for handling user authentication. Google Accounts,
+Salesforce, and Azure AD v2 are prominent examples out in the wild today. While there are many other types of mechanisms
+out there, this one immediately came to mind because of its ubiquity across major platforms. For solutions that don't
+implement OIDC, [dex][] has become a popular tool of choice that can convert many types of authentication flows into a
+provider. I've also seen [hydra][], which seems to allow companies to bolt and OIDC provider onto the side of an
+existing login workflow.
+
+In the end, making the decision to go with OIDC seemed like the right thing to do given its rather extensive use and our
+longer term vision of a more decentralized web.
+
+[dex]: https://github.com/dexidp/dex/
+
+[hydra]: https://github.com/ory/hydra
+
+## Implementation
+
+- Satellite Changes
+    - API / OIDCP
+        - New OAuth2 and OpenID endpoints to properly support authenticating users
+        - Database or Redis for token/code storage
+    - Console
+        - New workflow: Log in, MFA, Consent Screen
+        - New page: Consent Screen
+    - Admin
+        - New endpoint for being able to set up new clients (temporary)
+
+## Wrap up
+
+[Who will archive the blueprint when completed? What documentation needs to be updated to preserve the relevant information from the blueprint?]
+
+## Open issues
+
+[A discussion of issues relating to this proposal for which the author does not know the solution. This section may be omitted if there are none.]
+
+- How do we want to handle the problem with satellites?
+    - Keep them separate (bad developer experience) or merge them together (more work)
+    - There are definitely a handful of other options to consider as well. Looking to see if others have strong feelings
+      one way or another.
```

Comment:
One of the issues here is that we want community satellites as well and also "open source satellites" (that we haven't validated). In those cases it would be weird to have a central registry and provide a redirect mechanism on some site... but it would improve the devexp.

One approach to avoid this issue is to provide a "use a custom satellite provider" on the selection screen. But at that point the communication should probably happen directly with the custom satellite not through some of our proxy. I don't remember openid connect mechanism -- so I'm not sure how plausible that approach is.

Reply:
Its definitely not possible from the server perspective (or at least not reliable, etc). Based on the conversations I've had around this with folks it's starting to feel like, this is a "client must configure" type of thing (i.e. a responsibility of the 3rd party applications).

For v0 sake, it's probably worth scoping this out and having each satellite be different, but encourage app developers to support directly configuring in the browser. We could even give them an example form to follow.

Reply:
In OIDC, if we implement the discovery portion, then all third parties should have to prompt for is the base URL